executor/sort.go

```diff
@@ -412,3 +412,257 @@ func (e *TopNExec) Next(goCtx goctx.Context) (Row, error) {
+		e.totalCount += srcChk.NumRows()
+	}
+	e.initPointers()
+	e.initCompareFuncs()
+	allColumnExpr := e.buildKeyColumns()
+	if !allColumnExpr {
+		e.buildKeyExprsAndTypes()
+		err := e.buildKeyChunks()
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+const topNCompactionfactor = 4
+
+func (e *TopNExec) executeTopN() error {
+	heap.Init(e.chkHeap)
+	for len(e.rowPointers) > e.totalLimit {
+		// The number of rows we loaded may exceeds total limit, remove greatest rows by Pop.
+		heap.Pop(e.chkHeap)
+	}
+	var childKeyChk *chunk.Chunk
+	if len(e.keyChunks) > 0 {
+		childKeyChk = chunk.NewChunk(e.keyTypes)
+	}
+	childRowChk := e.children[0].newChunk()
+	for {
+		err := e.children[0].NextChunk(childRowChk)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if childRowChk.NumRows() == 0 {
+			break
+		}
+		err = e.processChildChk(childRowChk, childKeyChk)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if e.totalCount > len(e.rowPointers)*topNCompactionfactor {
+			err = e.doCompaction()
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
+	if len(e.keyChunks) != 0 {
+		sort.Slice(e.rowPointers, e.keyChunksLess)
+	} else {
+		sort.Slice(e.rowPointers, e.keyColumnsLess)
+	}
+	return nil
+}
+
+func (e *TopNExec) buildKeyExprsAndChildKeyChunk() ([]expression.Expression, *chunk.Chunk) {
+	keyLen := len(e.ByItems)
+	keyTypes := make([]*types.FieldType, keyLen)
+	keyExprs := make([]expression.Expression, keyLen)
+	for keyColIdx := range e.ByItems {
+		keyExpr := e.ByItems[keyColIdx].Expr
+		keyType := keyExpr.GetType()
+		keyExprs[keyColIdx] = keyExpr
+		keyTypes[keyColIdx] = keyType
+	}
+	return keyExprs, chunk.NewChunk(keyTypes)
+}
+
+func (e *TopNExec) processChildChk(childRowChk, childKeyChk *chunk.Chunk) error {
+	if childKeyChk != nil {
+		childKeyChk.Reset()
+		err := expression.VectorizedExecute(e.ctx, e.keyExprs, childRowChk, childKeyChk)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	for i := 0; i < childRowChk.NumRows(); i++ {
+		heapMaxPtr := e.rowPointers[0]
+		var heapMax, next chunk.Row
+		if childKeyChk != nil {
+			heapMax = e.keyChunks[heapMaxPtr.chkIdx].GetRow(int(heapMaxPtr.rowIdx))
+			next = childKeyChk.GetRow(i)
+		} else {
+			heapMax = e.rowChunks[heapMaxPtr.chkIdx].GetRow(int(heapMaxPtr.rowIdx))
+			next = childRowChk.GetRow(i)
+		}
+		if e.chkHeap.greaterRow(heapMax, next) {
+			// Evict heap max, keep the next row.
+			e.appendRowChunk(childRowChk.GetRow(i))
+			if childKeyChk != nil {
+				e.appendKeyChunk(childKeyChk.GetRow(i))
+			}
+			e.totalCount++
+			chkIdx := len(e.rowChunks) - 1
+			e.rowPointers[0].chkIdx = uint32(chkIdx)
+			e.rowPointers[0].rowIdx = uint32(e.rowChunks[chkIdx].NumRows() - 1)
+			heap.Fix(e.chkHeap, 0)
+		}
+	}
+	return nil
+}
+
+func (e *TopNExec) appendRowChunk(row chunk.Row) {
+	lastChkIdx := len(e.rowChunks) - 1
+	if lastChkIdx == -1 || e.rowChunks[lastChkIdx].NumRows() >= e.ctx.GetSessionVars().MaxChunkSize {
```

Comment:
this chunk is only used to buffer data, no need to be limited to `e.ctx.GetSessionVars().MaxChunkSize`

Reply:
But if the `offset + limit` is too large, store all rows in one chunk requires very large amount of contigous memory.

Reply:
how about ignore `e.ctx.GetSessionVars().MaxChunkSize` and limit the chunk size to 1024 ?